frontend/packages/knative-plugin/src/components/add/event-sources/KafkaSourceSection.tsx

```diff
@@ -1,43 +1,109 @@
 import * as React from 'react';
-import { useFormikContext, FormikValues } from 'formik';
+import * as _ from 'lodash';
 import FormSection from '@console/dev-console/src/components/import/section/FormSection';
-import { InputField, TextColumnField } from '@console/shared';
+import { InputField, SelectInputField, SelectInputOption } from '@console/shared';
 import { TextInputTypes } from '@patternfly/react-core';
 import KafkaSourceNetSection from './KafkaSourceNetSection';
 import ServiceAccountDropdown from '../../dropdowns/ServiceAccountDropdown';
+import { K8sResourceKind } from '@console/internal/module/k8s';
+import { useK8sWatchResources } from '@console/internal/components/utils/k8s-watch-hook';
+import { getBootstrapServers } from '../../../utils/create-eventsources-utils';
+import { strimziResourcesWatcher } from '../../../utils/get-knative-resources';
 
 const KafkaSourceSection: React.FC = () => {
-  const { values } = useFormikContext<FormikValues>();
-  const {
-    data: {
-      kafkasource: { bootstrapServers, topics },
-    },
-  } = values;
+  const [bootstrapServers, setBootstrapServers] = React.useState<SelectInputOption[]>([]);
+  const [bsPlaceholder, setBsPlaceholder] = React.useState<React.ReactNode>('');
+  const [kafkaTopics, setKafkaTopics] = React.useState<SelectInputOption[]>([]);
+  const [ktPlaceholder, setKtPlaceholder] = React.useState<React.ReactNode>('');
+
+  const memoResources = React.useMemo(() => strimziResourcesWatcher(), []);
+  const { kafkas, kafkatopics } = useK8sWatchResources<{
+    [key: string]: K8sResourceKind[];
+  }>(memoResources);
+
+  React.useMemo(() => {
```

Comment:
`useEffect` was fine for how you were doing because you were setting state in the effect.
When you use `useMemo` you should _not_ call any `setState` functions. But rather you should return the values.

eg.
```
const [ bootstrapServers, placeholder ] = React.useMemo(...
```

This way the values are calculated in place instead of round tripping through set state.

Reply:
I know I was missing something. Updating it now.

Reply:
Though it didn't stop me from using `setState` functions but yeah technically `useMemo`is used to get `memoized` values so no point in explicitly setting the state and then using it in fields.